### PR TITLE
Order processing

### PR DESCRIPTION
**Pull request by leodepinho**
Topics:

- Add order processing to the webhook endpoint.
- Add timestamp validation in webhook endpoint as a security measure.
- Update notification logic for Revolut Pay events.
- add template engine (`ejs`).
- add `REVOLUT_API_URL` as an environment variable.
- Fix typos.

Docs:

- https://webhooks.fyi/security/replay-prevention
- https://github.com/mcollina/fastq
- https://ejs.co/

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -118,24 +120,46 @@ app.post("/webhook", (req, res) => {
...
+    // For more information visit https://webhooks.fyi/security/replay-prevention
+    const isTimestampValid = validateTimestamp(revolutRequestTimestamp);
+
+    if (!isTimestampValid) {
+      // The request timestamp is outside of the tolerance zone.
+      return res.status(403).send("Timestamp outside the tolerance zone");
+    }
+
     if (isSignatureValid) {
+      const order = orders.getOrderByRevolutId(req.body["order_id"]);
```
Comment (andrewbents) [inline server/app.js:133]: in real world it would be a DB call, should we do it here before sending 200?

File: server/app.js [hunk 2]
```diff
@@ -118,24 +120,46 @@ app.post("/webhook", (req, res) => {
...
+
       // Return 200 response before doing any order computation
       res.sendStatus(200);
 
       switch (req.body.event) {
         // Add any order management logic (send emails, update orders, etc)
         case "ORDER_COMPLETED":
           console.log("Webhook - Order Completed!");
+          ordersQueue.push({
+            orderId: order.id,
```
Comment (andrewbents) [inline server/app.js:143]: is it the same order id as in `req.body["order_id"]`?

File: server/app.js [hunk 3]
```diff
@@ -118,24 +120,46 @@ app.post("/webhook", (req, res) => {
...
+    if (!isTimestampValid) {
+      // The request timestamp is outside of the tolerance zone.
+      return res.status(403).send("Timestamp outside the tolerance zone");
+    }
+
     if (isSignatureValid) {
+      const order = orders.getOrderByRevolutId(req.body["order_id"]);
+
       // Return 200 response before doing any order computation
       res.sendStatus(200);
```
Comment (andrewbents) [inline server/app.js:136]: now that there's an async queue, let's maybe schedule order handling and then return 200.
We can also add a comment above scheduling saying something like `Don't process orders directly in the webhook handler. Schedule order management logic (send emails, update orders, etc) and return 200 as soon as possible`

File: server/.env.example [hunk 4]
```diff
@@ -1,4 +1,5 @@
 STATIC_DIR=../client/html
+REVOLUT_API_URL=https://sandbox-merchant.revolut.com
```
Comment (andrewbents) [inline server/.env.example:2]: should we add some note in the README that merchants can try prod by changing the url and keys?

File: server/app.js (hunk 3 above)
Comment (andrewbents) [reply server/app.js:136]: we can even just push the event to the queue as is, and then process the order in the worker

File: server/app.js (hunk 1 above)
Comment (leodepinho) [reply server/app.js]: Makes sense, moved all order processing inside the worker

File: server/app.js (hunk 2 above)
Comment (leodepinho) [reply server/app.js]: No, it's not.  This one is the internal one (from merchant's) after the DB call:
`const order = orders.getOrderByRevolutId(req.body["order_id"]);`

But also moved that call inside of the worker.

File: server/app.js (hunk 3 above)
Comment (leodepinho) [reply server/app.js]: Agree, updated

File: server/.env.example (hunk 4 above)
Comment (leodepinho) [reply server/.env.example:2]: Updated

File: server/helpers.js [hunk 5]
```diff
@@ -18,3 +18,13 @@ export const validateSignature = ({
 
   return originalSignature === signature;
 };
+
+// 5 minutes tolerance (seconds)
+const TOLERANCE_ZONE = 300;
+
+export const validateTimestamp = (requestTimestamp) => {
+  const currentTimestamp = Math.floor(Date.now() / 1000);
+  const difference = currentTimestamp -  Math.floor(requestTimestamp / 1000);
```
Comment (andrewbents) [inline server/helpers.js:27]: do we need to divide them both by 1k? can we instead have the tolerance as `300_000`?

Comment (leodepinho) [reply server/helpers.js]: Yep I'll do it, normally milliseconds were not that important from what I saw in other implementations. Updated

**Review by andrewbents — APPROVED** [review]

**Pull request merged**